src/OrbitQt/HistogramWidget.cpp

```diff
@@ -87,74 +93,129 @@ static void DrawVerticalLine(QPainter& painter, const QPoint& start, int length)
   painter.drawLine(start, {start.x(), start.y() + length});
 }
 
+struct Ticks {
+  std::vector<QString> labels;
+  std::vector<double> values;
+  int precision{};
+};
+
+[[nodiscard]] static std::vector<QString> DoublesToLabels(const std::vector<double>& values,
+                                                          int precision) {
+  std::vector<QString> labels;
+  std::transform(
+      std::begin(values), std::end(values), std::back_inserter(labels),
+      [precision](const double value) { return QString::number(value, 'f', precision); });
+  return labels;
+}
+
+[[nodiscard]] static double Round(double x, int precision) {
+  const double powerOfTen = pow(10.0, precision);
+  return static_cast<double>(RoundToClosestInt(x * powerOfTen)) / powerOfTen;
+}
+
+[[nodiscard]] static std::vector<double> RoundDoubles(const std::vector<double>& values,
+                                                      int precision) {
+  std::vector<double> rounded;
+  std::transform(std::begin(values), std::end(values), std::back_inserter(rounded),
+                 [precision](const double value) { return Round(value, precision); });
+  return rounded;
+}
+
+template <typename T>
+bool AreAllUnique(std::vector<T> vector) {
+  std::sort(vector.begin(), vector.end());
+  return std::unique(vector.begin(), vector.end()) == vector.end();
+}
+
+[[nodiscard]] static Ticks GetTicklabels(const std::vector<double>& values, int max_precision) {
+  Ticks result;
+  for (int precision = 0; precision <= max_precision; precision++) {
+    auto labels = DoublesToLabels(values, precision);
+    if (precision == max_precision || AreAllUnique(labels)) {
+      auto rounded_values = RoundDoubles(values, precision);
+      result = {labels, rounded_values, precision};
+      break;
+    }
+  }
+  return result;
+}
+
+[[nodiscard]] static std::vector<double> GenerateEquidistantTickValues(double min, double spacing) {
+  std::vector<double> values;
+  double current = min;
+  for (uint32_t i = 0; i <= kHorizontalTickCount; ++i) {
+    values.push_back(current);
+    current += spacing;
+  }
+  return values;
+}
+
+[[nodiscard]] static Ticks GenerateLabels(double min, double spacing, int max_precision) {
+  return GetTicklabels(GenerateEquidistantTickValues(min, spacing), max_precision);
+}
+
+[[nodiscard]] static int ValueToAxisLocation(double value, int axis_length, double min_value,
+                                             double max_value) {
+  if (min_value == max_value) max_value++;
```

Comment:
making a double_variable++ is a bit strange to me. 

Can't you directly return here?

Reply:
I can. Good point.